lib/elixir/lib/code/formatter.ex

```diff
@@ -1016,6 +1016,7 @@ defmodule Code.Formatter do
   defp remote_to_algebra({target, meta, args}, context, state) do
     {target_doc, state} = quoted_to_algebra(target, :no_parens_arg, state)
 
+    state = Map.put_new(state, :ignore_empty_parens?, true)
```

Comment:
We avoid relying on state because it is often more complex, as we need to make sure to clean it up and so on.

For this particular issue, we already have a way to control how parens are emitted, which is exactly the `:required` atom below. You can see all available values here: https://github.com/elixir-lang/elixir/pull/9211/files#diff-2f18294c6d186e77413788b902525c57R1101

You can simply check for `:required` in the code below and emit "() do" accordingly. However, if you do so, two other tests will fail. That's because I think we need a third mode, which would be like `:skip_if_empty_do_end`, that only skips if we are in the "empty do end" part.
My suggestion is to introduce a new option, called

Reply:
Yeah, I figured `state` was a stupid idea but I couldn't come up with anything better. I didn't even realize `parens` was passed there. I adjusted the code to use `parens` instead.